Fixing invalid writes reported by valgrind for PP-746

suresh-thelkar opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* *Valgrind is reporting some of invalid writes which were caused by PP-746*

#### Affected Platform(s)
* *All platforms*

#### Cause / Analysis / Design
* This issue occurs when scheduler/s state is either "down" or "idle" in the database which can happen in the following corner cases only.
* Scheduler is shutdown abruptly and server is restarted
* Server is shutdown abruptly when scheduler/s state is "idle".

In other words we can say these invalid writes occur by restarting server only when state of scheduler/s in database is "idle" or "down".

#### Solution Description
* Do not save the state of scheduler/s in to the database. This was the original intention of PP-746 but somehow we missed this during that time.  Server itself can decide the state of scheduler/s, so need of saving this state in to the database. This itself will resolve the current issue.
* In addition to the above we are now using corresponding setter functions so that it correctly frees and allocates memory.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

suresh-thelkar commented:
No need to write separate PTL test cases for this. I have run all the multisched test cases under valgrind and got the clean report where invalid writes which caused due to PP-746 are absent now.

mike0042 commented on src/server/pbsd_init.c:
```diff
@@ -256,14 +256,18 @@ init_server_attrs()
-	struct attribute attrib = {0};
-	int i = 0;
+	attribute 	temp_attr;
+	struct 		attribute attrib = {0};
```
In attribute.h we have... typedef struct attribute attribute;
Normally, (for the sake of clarity) this would be... typedef struct attribute attribute_t;
You've used both "attribute" and "struct attribute" here, even though they are the same thing. Please use one or the other.

suresh-thelkar replied:
@mike0042,
I agree with you that we could have actually defined it as struct attribute attribute_t. I am using attribute only in both the places.
Since I have now created a separate function to set value for any attribute through server this temp_attr is not required here.(But defined inside the function set_attr_svr)

arungrover commented on src/server/pbsd_init.c:
```diff
@@ -256,14 +256,18 @@ init_server_attrs()
-	server.sv_attr[(int)SRV_ATR_scheduler_iteration].at_val.at_long =
-		PBS_SCHEDULE_CYCLE;
+	temp_attr.at_flags = ATR_VFLAG_SET;
+	temp_attr.at_val.at_long = PBS_SCHEDULE_CYCLE;
```
@suresh-thelkar Please use at_decode function to set the values here and in all other places because decode is a common way of populating an attribute and if that changes in future we will not have to worry about these special cases. It will also set the flag for you.
Also, if you end up using decode, don't forget to call at_free as well because that will then free up memory inside the attribute (if allocated)

suresh-thelkar replied:
Thanks for letting me know this. My initial idea was we should use decode functions only when we are getting data from external source i.e. either from database or from the user as input. One more thing what made me to use only set functions is most of the places in server code we are directly setting  value for an attribute of an object. But it seems it is better to use decode() followed by set functions as suggested by you. I have actually come up with a function named set_attr_svr() which takes care of all these and I am calling this wherever required.

And to correct all the places where we are not following this convention as of now which are outside the scope of this bug planning to raise an infrastructure RFE so that all the code we follow one approach. When we implement this we can actually enhance the newly written function set_attr_svr().

bhroam commented on src/server/svr_recov_db.c:
```diff
@@ -504,8 +504,10 @@ sched_save_db(pbs_sched *ps, int mode)
+		ps->sch_attr[SCHED_ATR_sched_state].at_flags &= ~(ATR_VFLAG_MODIFY | ATR_VFLAG_MODCACHE);
 		if (save_attr_db(conn, &attr_info, sched_attr_def, ps->sch_attr, (int)SCHED_ATR_LAST, flag) !=0)
 			goto db_err;
+		ps->sch_attr[SCHED_ATR_sched_state].at_flags &= (ATR_VFLAG_MODIFY | ATR_VFLAG_MODCACHE);
```
Did you mean |= ?  This will get rid of all bits other than modify and modcache

suresh-thelkar replied:
Implemented.

arungrover commented on src/server/run_sched.c:
```diff
@@ -359,9 +365,9 @@ schedule_high(pbs_sched *psched)
 	}
-	strncpy(psched->sch_attr[(int) SCHED_ATR_sched_state].at_val.at_str, SC_SCHEDULING, SC_STATUS_LEN);
-	psched->sch_attr[(int) SCHED_ATR_sched_state].at_val.at_str[SC_STATUS_LEN] = '\0';
-	psched->sch_attr[(int) SCHED_ATR_sched_state].at_flags = ATR_VFLAG_DEFLT | ATR_VFLAG_SET | ATR_VFLAG_MODCACHE;
```
You may still want to retain this line to add "ATR_VFLAG_DEFLT"  (in this place and all other places) because no setter function will add this for you.

suresh-thelkar replied:
It looks we are not really making use of this attribute in Multisched code and hence not using the flag.

bhroam commented on src/server/run_sched.c:
```diff
@@ -344,13 +344,19 @@ int
-		if ((s = contact_sched(psched->svr_do_sched_high, NULL, psched->pbs_scheduler_addr, psched->pbs_scheduler_port)) < 0)
+		if ((s = contact_sched(psched->svr_do_sched_high, NULL, psched->pbs_scheduler_addr, psched->pbs_scheduler_port)) < 0) {
+			temp_attr.at_val.at_str = SC_DOWN;
+			sched_attr_def[(int) SCHED_ATR_sched_state].at_set(&psched->sch_attr[(int) SCHED_ATR_sched_state], &temp_attr, SET);
```
set_str() will return an error if malloc() fails.  Do you want to catch that?  (for all other instances of setting the scheduler state as well).

suresh-thelkar replied:
I have actually written a function set_attr_svr() which takes care of decoding, setting and clearing the attribute. I am logging appropriate error messages if corresponding functions fail.

suresh-thelkar commented on src/server/pbsd_init.c:
```diff
@@ -256,14 +256,18 @@ init_server_attrs()
+	int 		i = 0;
+
 
 	for (i=0; i<SRV_ATR_LAST; i++)
```
Implemented.

suresh-thelkar commented on src/server/pbsd_init.c:
```diff
@@ -256,14 +256,18 @@ init_server_attrs()
-	server.sv_attr[(int)SRV_ATR_scheduler_iteration].at_val.at_long =
-		PBS_SCHEDULE_CYCLE;
+	temp_attr.at_flags = ATR_VFLAG_SET;
+	temp_attr.at_val.at_long = PBS_SCHEDULE_CYCLE;
```
Yes @bhroam, we need to do clear_attr() on temp_attr. Since I have now created a new function which is used to set a value for any attribute from server I am taking care of calling this in that function set_attr_svr().

suresh-thelkar commented:
@bhroam, @arungrover, @mike0042 
I have implemented all of your comments. Please look into the latest commit and provide your sign off or further review comments if any.

mike0042 commented on src/include/server.h:
```diff
@@ -253,6 +253,9 @@ enum failover_state {
 #define SVR_JOBHIST_DEFAULT	1209600	/* default time period to keep job history: 2 weeks */
 
+#define VALUE(str) #str
+#define TOSTR(str) VALUE(str)
```
Well, isn't that clever.  :-)

bhroam replied:
I think I'm missing something here.  Why do you need two macros?  Why can't you just have one macro that does #str?  According to my book, the # preprocessor will quote what follows.  I don't see what having the extra macro buys us.

suresh-thelkar replied:
I would like share you how I came to write such a macro. 

Following is the simple program that I tried initially.

#include <stdio.h>
#define PBS_SCHEDULE_CYCLE    600 /* re-schedule even if no change, 10 min   */

#define VALUE(str) #str
#define TOSTR(str) VALUE(str)

unsigned char s1[60] = TOSTR (PBS_SCHEDULE_CYCLE);    /* generates "600" \*/
unsigned char s2[60] = VALUE(PBS_SCHEDULE_CYCLE);     /* generates "PBS_SCHEDULE_CYCLE" \*/
int main()
{
        printf("s1 = %s\n",s1);
        printf("s2 = %s\n",s2);
        return 0;
}
Since we need the value of PBS_SCHEDULE_CYCLE as a string, I had to use both of the above macros.

mike0042 replied:
FWIW, it looks correct to me.

vstumpf replied:
In my opinion, I would flip them. `TOSTR` should return the name of the `#define` as a string, and `VALUE` should return the value of `#define`.

bhroam replied:
Thanks for the explanation.  I see it now.  I agree with @mike0042, this is quite clever.  As to @vstumpf's comment, the usable macro is TOSTR().  The other macro can be called anything.  I think TOSTR() correctly defines what is going on.

vstumpf replied:
Ah, I see now that TOSTR is making the value of a `#define` to a string. Makes sense to me.

mike0042 commented on src/server/svr_func.c:
```diff
@@ -7193,3 +7193,38 @@ action_opt_bf_fuzzy(attribute *pattr, void *pobj, int actmode)
+ *
+ */
+void
+set_attr_svr(attribute *pattr, attribute_def *pdef, char *value)
```
We already have functions named set_attribute, mgr_set_attr, set_attr, and set_attr_resc. The name set_attr_svr is innocuous, but not very descriptive. Besides the changes proposed in this PR, aren't there dozens of places where this should be called as opposed to the "old way" of setting server attributes?

If you're going to introduce a new function, I suggest you look for all the places it could be applied and refactor accordingly.

suresh-thelkar replied:
@mike0042,
I named the function as set_attr_svr because we can read it like "setting an attribute of any object from server" thinking that the older style of setting attributes is mostly used in server and hence named the function like that. Please suggest me a better name if you already have something in your mind. I am fine to change it.

As I mentioned earlier since I am going to log a separate infrastructure RFE to correct the code  where the older style is used(apparently it seems there are several places like that) to newer style we will revisit this function at that and modify it accordingly if required.

mike0042 replied:
I don't have a problem with the name you chose, it's just that we already have several functions with similar names. Every time we add another it makes it more difficult to determine which is the correct call to use in a given scenario. Please link the refactoring ticket to PP-746 so we can tell they are related.

suresh-thelkar replied:
Done

bhroam commented on src/server/pbsd_init.c:
```diff
@@ -256,17 +256,15 @@ init_server_attrs()
 	resource        *presc = NULL;
-	attribute 	temp_attr;
-	struct 		attribute attrib = {0};
+	attribute	attrib = {0};
```
I believe clear_attr() is the right answer.  It will memset() the structure and set some members to the right values.

suresh-thelkar replied:
Thanks @bhroam for answering it.  As we are now calling clear_attr() as part of the new function I have changed the definition of this variable back to attribute attrib; Removed the initialization here.

bhroam commented on src/server/svr_func.c:
```diff
@@ -7193,3 +7193,38 @@ action_opt_bf_fuzzy(attribute *pattr, void *pobj, int actmode)
+	attribute tempat;
+
+	if (pattr == NULL || pdef == NULL) {
+		(void)sprintf(log_buffer,  "Invalid pointer to attribute or its definition");
```
use snprintf() (and other uses in this function)

suresh-thelkar replied:
Done

bhroam commented on src/server/svr_func.c:
```diff
@@ -7193,3 +7193,38 @@ action_opt_bf_fuzzy(attribute *pattr, void *pobj, int actmode)
+
+	if (pattr == NULL || pdef == NULL) {
+		(void)sprintf(log_buffer,  "Invalid pointer to attribute or its definition");
+		log_err(errno, __func__, log_buffer);
```
errno will not be set.  You'll end up with a weird message which says "Success".  Maybe PBSE_INTERNAL?

suresh-thelkar replied:
I am now collecting the return value in rc both from decode() and set() functions and using the same.

mike0042 replied:
That may be a better choice than using -1 as I saw in a couple places.

suresh-thelkar commented:
@bhroam / @mike0042,
I have implemented your review comments. Please look into the latest commit and provide your sign off or further comments if any.
@arungrover,
Thanks for your sign offf. I have made couple of changes to implement Bhroam and Mike's review comments. Please go through the latest commit and provide your sign off or further review comments if any.

mike0042 commented on src/server/svr_func.c:
```diff
@@ -7212,19 +7212,20 @@ void
+		snprintf(log_buffer,  sizeof(log_buffer), "decode of %s failed", pdef->at_name);
+		log_err(rc, __func__, log_buffer);
+	} else if ((rc = pdef->at_set(pattr, &tempat, SET)) != 0) {
+		(void)snprintf(log_buffer, sizeof(log_buffer), "set of %s failed", pdef->at_name);
```
Please remove the cast to void for this call as well.

suresh-thelkar replied:
Done

suresh-thelkar commented:
@bhroam / @mike0042 ,
I have implemented all of your comments. Can you please provide your sign off if you do not have any further comments.

suresh-thelkar commented:
@mike0042 / @arungrover,
Thank you for your approvals. 
@bhroam,
Since I do not see any further comments from you since yesterday I am assuming your sign off and as this code fix is urgent  I have squashed the commits and rebased it with master. Hope this is fine.
